include/seqan3/range/view/async_buffer.hpp

```diff
@@ -0,0 +1,513 @@
+// -----------------------------------------------------------------------------------------------------
+// Copyright (c) 2006-2019, Knut Reinert & Freie Universität Berlin
+// Copyright (c) 2016-2019, Knut Reinert & MPI für molekulare Genetik
+// This file may be used, modified and/or redistributed under the terms of the 3-clause BSD-License
+// shipped with this file and also available at: https://github.com/seqan/seqan3/blob/master/LICENSE.md
+// -----------------------------------------------------------------------------------------------------
+
+/*!\file
+ * \author Hannes Hauswedell <hannes . hauswedell AT fu-berlin . de>
+ * \brief Provides seqan3::view::async_buffer.
+ */
+
+#pragma once
+
+#include <thread>
+
+#include <seqan3/core/type_traits/all.hpp>
+#include <seqan3/contrib/parallel/buffer_queue.hpp>
+#include <seqan3/range/shortcuts.hpp>
+#include <seqan3/range/view/detail.hpp>
+#include <seqan3/std/concepts>
+#include <seqan3/std/iterator>
+#include <seqan3/std/ranges>
+
+//-----------------------------------------------------------------------------
+// This is the path a value takes when using this view:
+//   urange
+// → async_buffer_view.buffer
+// → async_buffer_iterator.cached_value
+// → user
+//-----------------------------------------------------------------------------
+
+namespace seqan3::detail
+{
+
+/*!\brief Adds async_buffer behavior to the underlying range.
+ * \tparam urng_t The underlying range type.
+ * \implements std::ranges::InputRange
+ * \ingroup view
+ */
+template <std::ranges::View urng_t>
+class async_buffer_view : public std::ranges::view_interface<async_buffer_view<urng_t>>
+{
+private:
+
+    //!\brief The iterator type for the underlying range.
+    using urng_iterator_type = std::ranges::iterator_t<urng_t>;
+
+    //!\brief The underlying range.
+    urng_t urange;
+
+    //!\brief The buffer queue.
+    contrib::fixed_buffer_queue<value_type_t<urng_t>> buffer;
+
+    //!\brief Iterates over the underlying range and pushes into the buffer until at end or stopped.
+    void produce()
+    {
+        auto it = std::ranges::begin(urange);
+        auto e  = std::ranges::end(urange);
+
+        detail::spin_delay delay{};
+
+        while ((it != e) && (!stop_producer.load()))// TODO likely
```

Comment:
> So we stop reading if either we've read everything, or the view tells us to stop the producer thread, right?

Yes!

> Latter is needed if we do something lazily with the output of the async I/O and we want to stop before having read everything, correct?

It's needed when destructing the view, before all elements are moved into/through the buffer. Because the background thread is still running and we need to stop it.

It's possible to use wait_push instead which blocks when the queue is full. But the problem is that we get a dead-lock if 
  * the queue is full and produce is stuck in wait
  * nothing pops elements off
  * we destruct the view